### PR TITLE
Simplify RuboCop::AST::Traversal

### PR DESCRIPTION
**Pull request by bquorning**
For all nodes with children, simply iterate over all child nodes and call the appropriate `on_#{type}` method. Some children are not `Node` instances, and must necessarily be skipped - these children appear to be either `nil` or instances of `Symbol`.

TODO: See if performance is significantly better or worse.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [ ] Added tests.
* [ ] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by bquorning** [comment]
Hmm, running on a 3000 files repository got around 3% slower. I’ll have to do some profiling before this PR can be considered ok.

File: lib/rubocop/ast/traversal.rb [hunk 1]
```diff
@@ -14,170 +14,38 @@ def walk(node)
...
+      SOME_CHILD_NODES.each do |type|
         module_eval(<<-RUBY, __FILE__, __LINE__ + 1)
           def on_#{type}(node)
-            if (child = node.children[0])
-              send(:"on_\#{child.type}", child)
-            end
-          end
-        RUBY
-      end
+            node.children.each do |child|
```
Comment (Drenmi) [inline lib/rubocop/ast/traversal.rb]: You could try `#each_child_node` here instead of `#children` + `#each`. It should ignore all non-node children.

**Comment by bbatsov** [comment]
Any updates here?

**Comment by bbatsov** [comment]
I'm closing this PR due to no recent activity. Feel free to re-open it if you ever come back to it.

**Pull request closed without merging**